js/src/row/map.ts

```diff
@@ -0,0 +1,134 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import { Data } from '../data';
+import { DataType, Struct } from '../type';
+import { valueToString } from '../util/pretty';
+import { instance as getVisitor } from '../visitor/get';
+import { instance as setVisitor } from '../visitor/set';
+import { instance as indexOfVisitor } from '../visitor/indexof';
+import { instance as iteratorVisitor } from '../visitor/iterator';
+
+/** @ignore */ const kKeys = Symbol.for('keys');
+/** @ignore */ const kVals = Symbol.for('vals');
+
+export class MapRow<K extends DataType = any, V extends DataType = any> {
+    private [kKeys]: Data<K>;
+    private [kVals]: Data<V>;
+    constructor(slice: Data<Struct<{ key: K; value: V }>>) {
+        this[kKeys] = slice.children[0] as Data<K>;
+        this[kVals] = slice.children[1] as Data<V>;
+    }
+    [Symbol.iterator]() {
+        return new MapRowIterator(this[kKeys], this[kVals]);
+    }
+    public toArray() { return Object.values(this.toJSON()); }
```

Comment:
Wouldn't it be more efficient not to go through `toJSON` here?

Reply:
Absolutely would, yeah :smile: